columnflow/production/util.py

```diff
@@ -39,3 +42,151 @@ def attach_coffea_behavior(
     :return: Array with correct behavior attached for collections
     """
     return attach_coffea_behavior_fn(events, collections=collections)
+
+
+#
+# general awkward array functions
+#
+
+def ak_extract_fields(arr, fields, **kwargs):
+    """
+    Build an array containing only certain `fields` of an input array `arr`,
+    preserving behaviors.
+    """
+    # reattach behavior
+    if "behavior" not in kwargs:
+        kwargs["behavior"] = arr.behavior
+
+    return ak.zip(
+        {
+            field: getattr(arr, field)
+            for field in fields
+        },
+        **kwargs,
+    )
+
+
+#
+# functions for operating on lorentz vectors
+#
+
+_lv_base = partial(ak_extract_fields, behavior=coffea.nanoevents.methods.nanoaod.behavior)
+
+lv_xyzt = partial(_lv_base, fields=["x", "y", "z", "t"], with_name="LorentzVector")
+lv_xyzt.__doc__ = """Construct a `LorentzVectorArray` from an input array."""
+
+lv_mass = partial(_lv_base, fields=["pt", "eta", "phi", "mass"], with_name="PtEtaPhiMLorentzVector")
+lv_mass.__doc__ = """Construct a `PtEtaPhiMLorentzVectorArray` from an input array."""
+
+lv_energy = partial(_lv_base, fields=["pt", "eta", "phi", "energy"], with_name="PtEtaPhiELorentzVector")
+lv_energy.__doc__ = """Construct a `PtEtaPhiELorentzVectorArray` from an input array."""
+
+
+def lv_sum(lv_arrays):
+    """
+    Return the sum of identically-structured arrays containing Lorentz vectors.
+    """
+    # don't use `reduce` or list comprehensions
+    # to keep memory use as low as possible
+    tmp_lv_sum = None
+    for lv in lv_arrays:
+        if tmp_lv_sum is None:
+            tmp_lv_sum = lv
+        else:
+            tmp_lv_sum = tmp_lv_sum + lv
+
+    return tmp_lv_sum
```

Comment:
Just some code polishing.

```suggestion
    if not lv_arrays:
        raise ValueError("sum of empty input lorentz vector arrays impossible")
    return sum(lv_arrays[1:], lv_arrays[0])
```

Reply:
This is a good suggestion and originally I actually had something similar, however, a big disadvantage is that it does not work if `lv_arrays` is a generator expression. We have this use case in the `mttbar` analysis where the entries in `lv_arrays` take up a large amount of memory, and after some profiling tests it turned out that looping over the generator explicitly and performing the addition in-place performed better in terms of the memory footprint. (The downside of this is unfortunately more verbose code.)

Despite the above, I actually noticed that this function is not needed for this PR. I copied it over from the `mttbar` analysis along with the others without checking to see if it was actually used. Given this, we can probably remove this from the PR, unless you think it would be useful to keep it in as a sort of "memory-efficient" implementation of LV addition?